[sparse][wip] split out semi-structured sparsity subclasses

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116023

Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[sparse][wip] split out semi-structured sparsity subclasses

Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@alexsamardzic This is still pretty rough so no need to do a real review for now - mostly just checking to see if you have any concerns with splitting out `SparseSemiStructuredTensor`. 

I left the class in as a template to handle some shared stuff like shape checking + padding, and also to avoid breaking the scripts that we have out in public so far.

alexsamardzic commented:
Thanks!  I'm fine with these changes; in general, my intention is just to try to have functionality of these two back-ends for sparse semi-structured GEMM on par, as much as possible.

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][wip] split out semi-structured sparsity subclasses"


Summary:

This PR splits out SparseSemiStructuredTensor into
SparseSemiStructuredTensorCUTLASS and
SparseSemiStructuredTensorCUSPASRELT.

This is because of the following reasons:
* different sparse shape constraints for cuSPARSELT and CUTLASS
* cuSPARSELt specific metadata (alg_id, fuse_transpose) that we need for
  LLM inference.
* general readability

I decided on two subclasses rather than nested subclasses because

* torch.compile support for nested subclasses is an unknown and currently
  there's no way to access the nested attributes to flatten the tensor.

* Although we could put shared code like aten.linear and padding support in the general subclass,
  I think the resulting code is a bit hard to understand, especially how
  transpositions are handled, since both the outer and inner tensor will
  have a shape.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
